### DISCUSSION TITLE
Control order of relationships when creating model

### DISCUSSION THREAD

**Discussion by kokoshneta** [filamentphp/filament#6625] [Help] [answered]
I have a create/edit page for a number of different classes that all extend a base Publication class (the individual child classes represent monographs, journal issues, articles, etc.). Each type of publication is stored in a separate table, but their primary keys are _unique across all of these tables_.

The form definitions for the Filament resource classes for these child classes contain a number of blocks/groups that act on relationships. Here is a brief example from the BookResource class, showing the relationships `metadata` and `product` being used:

```php
return $form
  ->schema([
    MarkdownEditor::make('title'),
    MarkdownEditor::make('subtitle'),
    Fieldset::make('Metadata')
      ->relationship('metadata')
      ->schema([
        TextInput::make('width'),
        TextInput::make('height'),
        TextInput::make('weight'),
      ]),
    Fieldset::make('Product data')
      ->relationship('product')
      ->schema([
        TextInput::make('product_number'),
        Select::make('product_group_id')
          ->options([… options here …])
      ])
  ])
;
```

The relevant parts of the table definitions look like this:

```yaml
# books (journal_issues, articles, etc.)
- id
- …other columns

# products
- id
- publication_id
- publication_type
- …other columns

# metadata
- id
- publication_id  # foreign key on products.publication_id
- …other columns
```

This causes an issue when creating a new Book model in Filament: after the Book model is created, it first tries to save the `metadata` relationship and then the `product` relationship – but since there’s a foreign key from metadata to product, it fails on saving to the metadata table, because the corresponding product row has not yet been inserted.

In other words, I need a way to control the order in which the relationships are saved to the database, to ensure that the `product` relationship is always saved **before** the `metadata` relationship.

Is there a way to do this within Filament?

**Comment by danharrin** [comment, marked as answer]
I think the order they appear in the schema is the order they are saved, so maybe change your form structure?

**Reply by kokoshneta** [reply]
Unfortunately, I can’t really switch things around very much – it’s a very large edit page with everything grouped together to give the user an intuitive overview, and the product stuff belongs further down than the metadata stuff. Would it be enough to just add an empty group at the top that calls the relationship but doesn’t contain any actual schema?

**Reply by kokoshneta** [reply]
To answer my own question: Yes, that does seem to do the trick! Wonderful!

**Reply by danharrin** [reply]
Yeah... could be seen as a bit of a hack though. Maybe add a comment with it to explain why you did it.
